### PR TITLE
Document neutralize_variable

### PR DESCRIPTION
**Pull request by fpagnoux**
Connected to openfisca/openfisca-doc#43

I tried something new as an experiment: deprecating without removing a function.

My motivation is to avoid the fixed cost of publishing a major (updating all the dependencies) for something small where retro-compatibility is easy to implement.

The workflow I imagine would be wait for the next major bumps to remove it for good. Either we:
- Remove it in the next PR that brings a major bump. Doesn't respect the one PR = one goal principle.
- Publish the next major bump as a pre-release. Then all deprecated stuff can be removed right after.

To be discussed later with @MattiSG @cbenz, this is not a priority.

### CODE REVIEW

File: openfisca_core/taxbenefitsystems.py [hunk 1]
```diff
@@ -244,8 +244,23 @@ def get_column(self, column_name, check_existence = False):
...
+        A neutralized variable always returns its default value when being computed.
+
+        Trying to set inputs for a neutralized variable has no effect.
+        """
+        self.update_column(variable_name, neutralize_column(self.get_column(variable_name)))
+
     def neutralize_column(self, column_name):
-        self.update_column(column_name, neutralize_column(self.get_column(column_name)))
+        print(
+            u"Warning: the neutralize_column method has been renamed to neutralize_variable. "
```
Comment (fpagnoux) [inline openfisca_core/taxbenefitsystems.py]: I didn't use `log.warning` as it does't work well because there is no `handlers`:

```
No handlers could be found for logger "openfisca_core.taxbenefitsystems"
```

**Comment by MattiSG** [comment]
> The workflow I imagine would be wait for the next major bumps to remove it for good.

Perfect.

> Remove it in the next PR that brings a major bump. Doesn't respect the one PR = one goal principle.

Indeed. But the cost is of adding one commit that removes all stuff that's been marked as deprecated. It's not good, but we're getting closer to having internalised version constraints and discipline on publishing, at which point we will be able to design a proper releasing process. I like this intermediary step.

File: openfisca_core/taxbenefitsystems.py [hunk 2]
```diff
@@ -244,8 +244,23 @@ def get_column(self, column_name, check_existence = False):
...
+            u"neutralize_column has thus been deprecated and will be removed soon. "
```
Comment (MattiSG) [inline openfisca_core/taxbenefitsystems.py]: soon → in the next major version.

File: openfisca_core/taxbenefitsystems.py [hunk 3]
```diff
@@ -244,8 +244,23 @@ def get_column(self, column_name, check_existence = False):
     def update_column(self, column_name, new_column):
         self.column_by_name[column_name] = new_column
 
+    def neutralize_variable(self, variable_name):
+        """
+        Neutralizes an OpenFisca variable existing in the tax and benefit system.
+
+        A neutralized variable always returns its default value when being computed.
```
Comment (MattiSG) [inline openfisca_core/taxbenefitsystems.py]: > when being computed → when computed

File: openfisca_core/taxbenefitsystems.py [hunk 4]
```diff
@@ -244,8 +244,23 @@ def get_column(self, column_name, check_existence = False):
...
+
+        Trying to set inputs for a neutralized variable has no effect.
```
Comment (MattiSG) [inline openfisca_core/taxbenefitsystems.py]: Shouldn't it raise a warning? 😕

File: openfisca_core/taxbenefitsystems.py (hunk 1 above)
Comment (MattiSG) [inline openfisca_core/taxbenefitsystems.py]: It might be worth setting log handlers, because this should definitely be printed to `stderr`.

File: doc/requirements.txt [hunk 5]
```diff
@@ -1,4 +1,4 @@
 # pip doesn't handle well dependencies with a relative path inside a requirements.txt file.
 # If we put .. instead of 'openfisca_core >= 4.2.0, < 5.0', it will look for a setup.py in the parent directory relatively to where the instruction is executed from, and not where the requirements file is.
-openfisca_core >= 4.2.0, < 6.0
+openfisca_core >= 9.0.0
```
Comment (MattiSG) [inline doc/requirements.txt]: I don't understand why this is changed.

File: openfisca_core/taxbenefitsystems.py (hunk 1 above)
Comment (cbenz) [inline openfisca_core/taxbenefitsystems.py]: I can help setting those logging handlers.

Handlers are set at the application / script (OpenFisca-Web-API) level, not the library (OpenFisca-Core) level. But the library can provide default settings.

Comment (fpagnoux) [inline openfisca_core/taxbenefitsystems.py]: The typical user of `neutralize_column ` does't use the web API, so we cannot rely on the Web API handlers. 
If you know how to set relevant default handlers, feel free to add a commit to this PR ;)

File: doc/requirements.txt (hunk 5 above)
Comment (fpagnoux) [inline doc/requirements.txt]: This is the version to be used in the [reference doc](http://openfisca.readthedocs.io/en/latest/). We want to take the last published.

**Comment by fpagnoux** [comment]
We should probably use [`warnings.warn`](https://docs.python.org/2/library/warnings.html#warnings.warn) and not bother with logging for this PR.

File: openfisca_core/taxbenefitsystems.py [hunk 6]
```diff
@@ -244,8 +245,24 @@ def get_column(self, column_name, check_existence = False):
     def update_column(self, column_name, new_column):
         self.column_by_name[column_name] = new_column
 
+    def neutralize_variable(self, variable_name):
+        """
+        Neutralizes an OpenFisca variable existing in the tax and benefit system.
+
+        A neutralized variable always returns its default value when computed.
+
+        Trying to set inputs for a neutralized variable has no effect.
```
Comment (cbenz) [inline openfisca_core/taxbenefitsystems.py]: > `Trying to set inputs` => `Setting inputs` ?

File: openfisca_core/taxbenefitsystems.py [hunk 7]
```diff
@@ -244,8 +245,24 @@ def get_column(self, column_name, check_existence = False):
...
+        """
+        self.update_column(variable_name, neutralize_column(self.get_column(variable_name)))
```
Comment (cbenz) [inline openfisca_core/taxbenefitsystems.py]: Should this helper `neutralize_column` be renamed to `neutralized_variable` (I insist on *neutralized*, like in `new_filled_column`)?

Because `TaxBenefitSystem.neutralize_variable` is an action, whereas the helper in `formulas.py` is a function which creates a new `Variable`.

**Comment by cbenz** [comment]
I added a commit to use [Python warning framework (PEP 230)](https://www.python.org/dev/peps/pep-0230/).

By default since Python 2.7, deprecation warnings are hidden by default ([explanation here](https://docs.python.org/2/library/warnings.html#updating-code-for-new-versions-of-python)).

The recommended way to display them is either:
- using the [`python -W` option](https://docs.python.org/2/using/cmdline.html#cmdoption-W)
- using the [`PYTHONWARNINGS`](https://docs.python.org/2/using/cmdline.html#envvar-PYTHONWARNINGS) environment variable

In short it is possible to convert deprecation warnings into exceptions (`python -W error`) or just display them (`python -W default`). This will have to be documented in a section of the doc with a title like "How to upgrade my OpenFisca user code" (any thought @fpagnoux @MattiSG)?

See also the "Command Line Syntax " section of the [PEP 230](https://www.python.org/dev/peps/pep-0230/) (sorry there is no direct link to an anchor).

**Comment by fpagnoux** [comment]
Thanks @cbenz for all the references, this is interesting. 

In our case, I don't think a `Warning` that is hidden by default is much useful. First, we cannot convince our users to always run `python` with the `-w` option.  Second, a "How to upgrade my OpenFisca user code" tutorial would definitely have some value, but it's a project on its own, and if it only contains: "Use python with -W", it's not fulfilling its promise.

As we are still in the experimental phase about deprecation, I would just add general `Warning` (visible by default). This is less semantic, but closer to the concrete behaviour we expect.

File: openfisca_core/taxbenefitsystems.py (hunk 6 above)
Comment (fpagnoux) [inline openfisca_core/taxbenefitsystems.py]: As we _cannot_ set an input for a neutralized variable, I prefer using `try`.

File: openfisca_core/taxbenefitsystems.py (hunk 7 above)
Comment (fpagnoux) [inline openfisca_core/taxbenefitsystems.py]: 👍
